### PR TITLE
Feature: Digit separators in numeric literals

### PR DESCRIPTION
**Pull request by Maxxen**
This PR is an updated continuation/resurrection of #5338

This PR makes it possible to include underscores ("_") as optional separators in numeric literals.
- Underscores are allowed in integer, decimal, hexadecimal and binary notation
- Underscores can not be the first or last character in a literal
- Underscores have to have an integer/numeric part on either side of them. I.e. there can not be multiple underscores in a row and not immediately before/after a decimal or exponent. 

The fact that multiple underscores can't follow each other is consistent with the SQL 2023 standard. Sadly I don't have a copy at hand and ISO wants way to much money to purchase one, but [here's a blog post](https://www.cybertec-postgresql.com/en/underscores-in-numeric-constants-in-postgresql-16/) describing the feature in postgresql

Some examples:

```sql

SELECT 100_000_000
----
100000000

SELECT 1_2.1_2E0_1 ;
----
121.2

SELECT '0xFF_FF'::INT;
----
65535

SELECT '0b0_1_0_1'::INT
----
 5
 
 # Although the following does not work:

SELECT '1__000__000'::INT;
----
Error: Conversion Error: Could not convert string '1__000__000' to INT32

SELECT '0xFF__FF'::INT;
 ----
Error: Conversion Error: Could not convert string '0xFF__FF' to INT32


# As these literals wont be recognized as numeric 
# the underscores get parsed as column aliases/references instead.

SELECT _1.0;
----
Error: Parser Error: syntax error at or near ".0"
LINE 1: SELECT _1.0;


SELECT 1.0_;
┌──────────────┐
│      _       │
│ decimal(2,1) │
├──────────────┤
│          1.0 │
└──────────────┘


```

### CODE REVIEW

**Comment by Maxxen** [comment]
Ah, still need to fix the decimal widening issue. Hold on.

File: test/sql/parser/digit_separators.test [hunk 1]
```diff
@@ -0,0 +1,314 @@
+# name: test/sql/parser/digit_separators.test
+# group: [parser]
+
+statement ok
+PRAGMA enable_verification
+
+# TODO: We should not allow multiple underscores:
```
Comment (Mytherin) [inline test/sql/parser/digit_separators.test:7]: This is implemented right?

**Review by Mytherin — APPROVED** [review]
Thanks! Looks great. Could you perhaps merge with main to see if that resolves the CI issues?

File: test/sql/parser/digit_separators.test (hunk 1 above)
Comment (Maxxen) [reply test/sql/parser/digit_separators.test]: Yes, removed comment

**Comment by Mytherin** [comment]
Thanks!

**Comment by szarnyasg** [comment]
This is now documented at https://duckdb.org/docs/sql/keywords-and-identifiers#numeric-literals

**Pull request merged**